### DISCUSSION TITLE
The reuse of optimal hyperparameters

### DISCUSSION THREAD

**Discussion by qweajl123** [optuna/optuna#5336] [Q&A]
### Issue Description
I've been working on optimizing hyperparameters for a neural network using the optuna The process involves two separate Python scripts: one for hyperparameter optimization and another for training the model with the optimized hyperparameters. Despite using identical datasets, model architectures, and ensuring the same random seed across both scripts, I observed a significant discrepancy in the training and validation losses between the optimization and training phases.

The hyperparameter optimization script successfully identifies what appears to be an optimal set of hyperparameters. However, when these parameters are manually transferred to the training script and the model is retrained, the resulting losses significantly differ from those observed during the hyperparameters optimization phase.

### Code1: Hyperparameters optimization

```

from Pensim_ODEmodel import Pensim
from neuromancer.dataset import DictDataset
from neuromancer.system import Node, System
from neuromancer.modules import blocks
from neuromancer.dynamics import integrators
import torch
from neuromancer.constraint import variable
from neuromancer.problem import Problem
from neuromancer.loss import PenaltyLoss
from L1_new_trainer import Trainer
from torch.utils.data import DataLoader
import matplotlib.pyplot as plt
import numpy as np
import dill
import optuna
import plotly
from optuna.visualization import plot_contour
from optuna.visualization import plot_edf
from optuna.visualization import plot_intermediate_values
from optuna.visualization import plot_optimization_history
from optuna.visualization import plot_parallel_coordinate
from optuna.visualization import plot_param_importances
from optuna.visualization import plot_slice
from optuna import samplers


def setup_seed(seed):
   torch.manual_seed(seed)
   torch.cuda.manual_seed_all(seed)
   np.random.seed(seed)
   torch.backends.cudnn.deterministic = True

setup_seed(55)

sys = Pensim()
nsim = 4000
nsteps = 10
nbatch = nsim//nsteps
length = (nsim//nsteps) * nsteps
ts = sys.ts
print(ts)
nx = sys.nx
nu = sys.nu


# data
with open('./train_dataset_save.pkl','rb') as f:
    train_data = dill.load(f)

with open('./dev_dataset_save.pkl','rb') as f:
    dev_data = dill.load(f)

with open('./test_save.pkl','rb') as f:
    test_data = dill.load(f)



def objective(trial):
    bs = trial.suggest_int('bs', 50, 200)
    q1 = trial.suggest_float('q1', 0.1, 20.0)
    q2 = trial.suggest_float('q2', 0.1, 20.0)
    nonlin_name = trial.suggest_categorical('nonlin', ['ReLU', 'SELU', 'Hardtanh', 'Tanh', 'ReLU6', 'GELU'])
    nonlin = getattr(torch.nn.modules.activation, nonlin_name)
    lr = trial.suggest_float('lr', 10**(-3.5), 10**(-1.5), log=True)
    weight_decay = trial.suggest_float('weight_decay', 1.e-6, 1.e-2, log=True)
    patience = trial.suggest_int('patience', 50, 300)
    epochs = trial.suggest_int('epochs', 300, 2000)
    warmup = trial.suggest_int('warmup', 50, 300)
    lambda_L1 = trial.suggest_float('lambda_L1', 1.e-6, 1.e-2, log=True)

    nonlin = getattr(torch.nn.modules.activation, nonlin_name)

    
    n_layers = trial.suggest_int('n_layers', 2, 3)

    
    hsizes = []
    for i in range(n_layers):
        
        n_units = trial.suggest_int(f'n_units_l{i}', 16, 128)
        hsizes.append(n_units)

    train_loader = DataLoader(train_data, batch_size=bs,
                              collate_fn=train_data.collate_fn, shuffle=True)

    dev_loader = DataLoader(dev_data, batch_size=bs,
                            collate_fn=dev_data.collate_fn, shuffle=True)

    
    fx = blocks.MLPDropout(nx + nu, nx, bias=True,
                           linear_map=torch.nn.Linear,
                           nonlin=nonlin,
                           hsizes=hsizes)

    fxRK4 = integrators.RK4(fx, h=ts)  

    system_nodel = Node(fxRK4, ['xn', 'U'], ['xn'], name='NODE')
    model = System([system_nodel])

    # Define the loss function

    # %% Constraints + losses:
    x = variable("X")  # real state
    xhat = variable('xn')[:, :-1, :]  # forecast state

    # trajectory tracking loss
    reference_loss = q1 * (xhat == x) ^ 2
    reference_loss.name = "ref_loss"

    onestep_loss = q2 * (xhat[:, 1, :] == x[:, 1, :]) ^ 2
    onestep_loss.name = "onestep_loss"

    # aggregate list of objective terms and constraints
    objectives = [reference_loss, onestep_loss]
    constraints = []
    # create constrained optimization loss
    loss = PenaltyLoss(objectives, constraints)
    # construct constrained optimization problem
    problem = Problem([model], loss)

    
    opt = torch.optim.Adam(problem.parameters(),
                           lr=lr, weight_decay=weight_decay)
    trainer = Trainer(
        problem,
        train_loader,
        dev_loader,
        test_data,
        opt,
        patience=patience,
        warmup=warmup,
        epochs=epochs,
        eval_metric="dev_loss",
        train_metric="train_loss",
        dev_metric="dev_loss",
        test_metric="dev_loss",
        lr_scheduler=True,
        # logger=logger,
        lambda_L1=lambda_L1
    )
    best_model = trainer.train()
    dev_loss = trainer.dev_losses[-1]
    train_loss = trainer.train_losses[-1]
    total_loss = dev_loss + train_loss

    with open('./trial/{}.pkl'.format(trial.number), 'wb') as f:
        dill.dump(best_model, f)

    return total_loss





study = optuna.create_study(
    direction="minimize",
    sampler=optuna.samplers.RandomSampler(seed=55),
    pruner=optuna.pruners.MedianPruner(n_warmup_steps=10),
)
study.optimize(objective, n_trials=10)  


best_params = study.best_params
best_trial = study.best_trials
print("Optimal hyperparameters：", best_params)
print(best_trial)

history = optuna.visualization.plot_optimization_history(study)
importance = optuna.visualization.plot_param_importances(study)
# plotly.offline.plot(history, importance, values, coordinate, contour, edf)
plotly.offline.plot(importance)
plotly.offline.plot(history)
```

### Code2: Re-train with the optimal hyperparameters
```

from Pensim_ODEmodel import Pensim
from neuromancer.dataset import DictDataset
from neuromancer.system import Node, System
from neuromancer.modules import blocks
from neuromancer.dynamics import integrators
import torch
from neuromancer.constraint import variable
from neuromancer.problem import Problem
from neuromancer.loss import PenaltyLoss
from L1_new_trainer import Trainer
from torch.utils.data import DataLoader
import matplotlib.pyplot as plt
import numpy as np
import dill

# seed
def setup_seed(seed):
   torch.manual_seed(seed)
   torch.cuda.manual_seed_all(seed)
   np.random.seed(seed)
   torch.backends.cudnn.deterministic = True

setup_seed(55)


sys = Pensim()
nsim = 4000
bs = 88   # batch size
nsteps = 10
nbatch = nsim//nsteps
length = (nsim//nsteps) * nsteps
ts = sys.ts
print(ts)
nx = sys.nx
nu = sys.nu

with open('./train_loader_save.pkl','rb') as f:
    train_loader = dill.load(f)

with open('./dev_loader_save.pkl','rb') as f:
    dev_loader = dill.load(f)

with open('./test_save.pkl','rb') as f:
    test_data = dill.load(f)




fx = blocks.MLPDropout(nx+nu, nx, bias=True,
                 linear_map=torch.nn.Linear,
                 nonlin=torch.nn.Hardtanh,
                 hsizes=[2,54,38])

fxRK4 = integrators.RK4(fx, h=ts)   

system_nodel = Node(fxRK4, ['xn', 'U'], ['xn'], name='NODE')
model = System([system_nodel])

# Loss function

# %% Constraints + losses:
x = variable("X")   # real state
xhat = variable('xn')[:, :-1, :]  # forecast state

# trajectory tracking loss
reference_loss = 4.898979947382423*(xhat == x)^2
reference_loss.name = "ref_loss"

onestep_loss = 18.207470630875203*(xhat[:, 1, :] == x[:, 1, :])^2
onestep_loss.name = "onestep_loss"

# aggregate list of objective terms and constraints
objectives = [reference_loss, onestep_loss]
constraints = []
# create constrained optimization loss
loss = PenaltyLoss(objectives, constraints)
# construct constrained optimization problem
problem = Problem([model], loss)


opt = torch.optim.Adam(problem.parameters(),
                             lr=0.015496696405929027, weight_decay=0.0003489569050625787)
trainer = Trainer(
    problem,
    train_loader,
    dev_loader,
    test_data,
    opt,
    patience=204,
    warmup=209,
    epochs=1674,
    eval_metric="dev_loss",
    train_metric="train_loss",
    dev_metric="dev_loss",
    test_metric="dev_loss",
    lr_scheduler=True,
    # logger=logger,
    lambda_L1=1.1740329265143504e-05
)
best_model = trainer.train()
problem.load_state_dict(best_model)





# update the rollout length based on the test data
model.nsteps = test_data['X'].shape[1]

# Test set results
test_outputs = model(test_data)

pred_traj = test_outputs['xn'][:, :-1, :].detach().numpy().reshape(-1, nx)
true_traj = test_data['X'].detach().numpy().reshape(-1, nx)
input_traj = test_data['U'].detach().numpy().reshape(-1, nu)
pred_traj, true_traj = pred_traj.transpose(1, 0), true_traj.transpose(1, 0)


plt.figure()
plt.plot(trainer.train_losses, label='Train Loss')
plt.plot(trainer.dev_losses, label='Eval Loss')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.title('Training Loss Over Epochs')
plt.legend()

figsize = 20
for k, (t1, t2) in enumerate(zip(true_traj, pred_traj)):
    plt.figure(figsize=(figsize, 6)) 
    plt.plot(t1, 'c', linewidth=4.0, label='True')  
    plt.plot(t2, 'm--', linewidth=4.0, label='Pred')  
    plt.ylabel(f'$y_{k}$', rotation=0, labelpad=20, fontsize=figsize)
    plt.tick_params(labelsize=figsize)  
    plt.legend(fontsize=figsize)  
    plt.title(f'Trajectory Comparison {k}')  

plt.show()
```

### Code 1 results in finding the optimal hyperparameter  (The optimal hyperparameter obtained in Trial 2)
![image](https://github.com/optuna/optuna/assets/86866346/a5c9e2fc-7916-438f-ac05-c070de2a89d0)

### Code 2  the loss with the optimal hyperparameter
![image](https://github.com/optuna/optuna/assets/86866346/7ca45c04-4b23-4472-a09c-e6c7bc21e926)

**Comment by nzw0301** [comment]
I'm not  core dev team, but thank you for your kind words. Personally, such words encouraging me to contributes optuna.

At the first gralnce, I think the current code uses different random states in the hyper-parameter optimisation and re-training steps. Concretely, to fix randomness, `setup_seed` should be called at every objective.
